src/wast-parser.cc

```diff
@@ -0,0 +1,2013 @@
+      func = dyn_cast<FuncModuleField>(&field)->func;
+    } else if (field.type == ModuleFieldType::Import) {
+      Import* import = dyn_cast<ImportModuleField>(&field)->import;
+      if (import->kind == ExternalKind::Func) {
+        func = import->func;
+      } else {
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    // Resolve func type variables where the signature was not specified
+    // explicitly, e.g.: (func (type 1) ...)
+    if (func->decl.has_func_type && IsEmptySignature(&func->decl.sig)) {
+      FuncType* func_type = module->GetFuncType(func->decl.type_var);
+      if (func_type) {
+        func->decl.sig = func_type->sig;
+      }
+    }
+
+    // Resolve implicitly defined function types, e.g.: (func (param i32) ...)
+    if (!func->decl.has_func_type) {
+      Index func_type_index = module->GetFuncTypeIndex(func->decl.sig);
+      if (func_type_index == kInvalidIndex) {
+        auto func_type = new FuncType();
+        func_type->sig = func->decl.sig;
+        module->AppendField(new FuncTypeModuleField(func_type, field.loc));
+      }
+    }
+  }
+}
+
+void AppendInlineExportFields(Module* module,
+                              ModuleFieldList* fields,
+                              Index index) {
+  Location last_field_loc = module->fields.back().loc;
+
+  for (const ModuleField& field: *fields) {
+    auto* export_field = cast<ExportModuleField>(&field);
+    export_field->export_->var = Var(index, last_field_loc);
+  }
+
+  module->AppendFields(fields);
+}
+
+}  // End of anonymous namespace
+
+WastParser::WastParser(WastLexer* lexer,
+                       ErrorHandler* error_handler,
+                       WastParseOptions* options)
+    : lexer_(lexer), error_handler_(error_handler), options_(options) {}
+
+void WastParser::Error(Location loc, const char* format, ...) {
+  errors_++;
+  va_list args;
+  va_start(args, format);
+  WastFormatError(error_handler_, &loc, lexer_, format, args);
+  va_end(args);
+}
+
+Token WastParser::GetToken() {
+  if (tokens_.empty())
+    tokens_.push_back(lexer_->GetToken(this));
+  return tokens_.front();
+}
+
+Location WastParser::GetLocation() {
+  return GetToken().loc;
+}
+
+TokenType WastParser::Peek() {
+  return GetToken().token_type;
+}
+
+TokenType WastParser::PeekAfter() {
+  if (tokens_.empty())
+    tokens_.push_back(lexer_->GetToken(this));
+  if (tokens_.size() < 2)
+    tokens_.push_back(lexer_->GetToken(this));
+
+  return tokens_.at(1).token_type;
+}
+
+bool WastParser::PeekMatch(TokenType type) {
+  return Peek() == type;
+}
+
+bool WastParser::PeekMatchLpar(TokenType type) {
+  return Peek() == TokenType::Lpar && PeekAfter() == type;
+}
+
+bool WastParser::PeekMatchExpr() {
+  return IsExpr(Peek(), PeekAfter());
+}
+
+bool WastParser::Match(TokenType type) {
+  if (PeekMatch(type)) {
+    Consume();
+    return true;
+  }
+  return false;
+}
+
+bool WastParser::MatchLpar(TokenType type) {
+  if (PeekMatchLpar(type)) {
```

Comment:
That's what `PeekMatchLpar` does, but it also checks that `Peek()` is `TokenType::Lpar`